### PR TITLE
recyclering

### PR DESCRIPTION
**Pull request by aeyakovenko**
Moving everything over to @garious recycler.  I am stuck on window, it seems to really want to clone those blobs.

### CODE REVIEW

File: src/recycler.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+/// `allocate()` is called, the value will be pulled from the
+/// recycler instead being allocated from memory.
+
+pub struct Recycler<T: Default + Reset> {
+    landfill: Arc<Mutex<Vec<T>>>,
+}
+impl<T: Default + Reset> Clone for Recycler<T> {
+    fn clone(&self) -> Self {
+        Recycler {
+            landfill: self.landfill.clone(),
```
Comment (rob-solana) [inline src/recycler.rs]: why do you have to clone the landfill, which is an Arc?

File: src/recycler.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+impl<T: Default + Reset> Default for Recycler<T> {
+    fn default() -> Self {
+        Recycler {
+            landfill: Arc::new(Mutex::new(vec![])),
+        }
+    }
+}
+
+impl<T: Default + Reset> Recycler<T> {
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/recycler.rs]: @rob-solana yea, same here.  this was @garious's clever tricks.  I think Recyclable needs to have an Arc to the item and a WeakRef to the landfill.

[the remark this answers is not in the document]
Comment (garious) [reply src/recycler.rs]: That's the big open question. I think the choice will be between switching all the stages to scoped threads or using Weak refs.  The former would be quite a bit cleaner and maps more directly to our architecture. Fullnode or TPU could host the recycler, and stages would reference it, and could be certain that they'll fall out of scope before the Fullnode/Tpu is dropped.

**Comment by garious** [comment]
I wouldn't be surprised if you continued to Arc the recyclables. But the recycler shouldn't have to be aware that you're doing that.  I'm crossing my fingers you don't need to Arc the recycler as well, because the recycler *would* have to be aware of that, and you'd therefore need to switch to the Weak ref trick.  It's all about proving the lifetime of the stage thread is shorter than the lifetime of the struct that holds its JoinHandle.  Scoped threads looks like the right tool for the job.

File: src/recycler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rob-solana) [reply src/recycler.rs]: Recycleable could be an Arc, too?  One AddRef() per copy instead of 2, as clone() would require copying an Arc and a Weak, right?

**Comment by rob-solana** [comment]
why would the Recycler need to be aware of Arc'ing?

File: src/recycler.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            .lock()
+            .unwrap()
+            .pop()
+            .map(|mut val| {
+                val.reset();
+                val
+            }).unwrap_or_default();
+        Recyclable {
+            val: Some(val),
+            landfill: self.landfill.clone(),
```
Comment (rob-solana) [inline src/recycler.rs]: if landfill is an Arc, do we need to clone() it here?

File: src/recycler.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+    pub fn recycle(&self, r: Recyclable<T>, _name: &str) {
```
Comment (rob-solana) [inline src/recycler.rs]: recycle() can take self instead of &self, instead of calling drop()?

Comment (rob-solana) [inline src/recycler.rs]: who calls this function?

Comment (rob-solana) [reply src/recycler.rs]: or...  why does anyone call this?

File: src/crdt.rs [hunk 5]
```diff
@@ -870,7 +870,7 @@ impl Crdt {
         blob_recycler: &BlobRecycler,
     ) -> Option<SharedBlob> {
         let pos = (ix as usize) % window.read().unwrap().len();
-        if let Some(blob) = &window.read().unwrap()[pos].data {
+        if let Some(ref mut blob) = &mut window.write().unwrap()[pos].data {
```
Comment (rob-solana) [inline src/crdt.rs]: this write() makes the window more of a Mutex<> than a RwLock<>

the as_mut() stuff is better than the RwLock<> stuff that was there before?

**Comment by aeyakovenko** [comment]
@garious @rob-solana i think this is an improvement.  there may be other tricks to make it simpler, but because the blobs are stored in the window concurrently with other stages we need the arc inside the recycler so it can be cloned.

**Comment by garious** [comment]
@aeyakovenko, my hope was that you could `impl Arc<Blob> for Recyclable` instead of moving that Arc into the recycler. Was that not possible?

File: src/recycler.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    fn drop(&mut self) {
+        if Arc::strong_count(&self.val) == 1 {
+            // this isn't thread safe, it will allow some concurrent drops to leak and not recycle
+            // if that happens the allocator will end up allocating from the heap
+            self.landfill.lock().unwrap().push(self.val.clone());
+        }
+    }
+}
+
+impl<T: Default + Reset> Clone for Recyclable<T> {
```
Comment (garious) [inline src/recycler.rs]: Seems like using this is an indication of a bug.

**Comment by garious** [comment]
I have a feeling the only Arc that's really needed is the one on Blobs so that it can be shared between the window and replicate stage.

**Comment by garious** [comment]
@aeyakovenko, I'll have some time later tonight to play with recyclables across threads (and scoped threads). While this PR is a big improvement over current code, I'd really like to get rid of the full class of recycler-related bugs here.

**Comment by aeyakovenko** [comment]
@garious Arc<Recyclable> might be doable.  maybe next pr.  this one touches tons of files, id like to land it since it seems like a huge improvement already.

File: src/recycler.rs (hunk 6 above)
Comment (aeyakovenko) [reply src/recycler.rs]: the window code... it keeps a copy of the blobs while they are referenced in other places.

**Comment by aeyakovenko** [comment]
Woot!  Merge it!

**Comment by garious** [comment]
Let me do my analysis first. I don't want anyone debugging ref count cycles if getting those Arcs out of there is easy enough.

**Comment by garious** [comment]
@aeyakovenko, I added 4 new tests to my original recycler (one of which you wrote). It shows that if you change our threads to crossbeam's scoped threads that you can use the recycler across threads with no Arc on the recycler and no Arcs on the recyclables.

https://github.com/garious/recycler/blob/master/src/recycler.rs#L165

**Comment by aeyakovenko** [comment]
@garious, great!  Let’s experiment with scoped threads in another Pr. This one is a monster that removes a bunch of recycler cruft already

**Comment by garious** [comment]
Can you implement your recycler in terms of mine so that at least there's a clear path to getting off the Arcs?

Maybe something like:

```rust
type ArcRecycler = Arc<Recycler<Arc<RwLock<T>>>;

trait ArcRecyclerUtil {
    fn read(&self) -> RwLockReadGuard;
    fn write(&self) -> RwLockWriteGuard;
}

impl ArcRecyclerUtil for ArcRecycler {
    ...
}
```

**Comment by aeyakovenko** [comment]
This change is to big to iterate in

**Comment by garious** [comment]
Yep, turned out to be the wrong approach. You should have either implemented only Drop for the existing recycler to remove `recycle()` calls or seen that the new recycler couldn't be integrated before moving to scoped threads. Instead you tossed in unrelated cleanup to make the PR so big that it's now in a take it or leave it state.

**Comment by aeyakovenko** [comment]
@garious, in what way does this make the current code base worse?  It doesn’t matter if it’s perfect, you can improve on this latter.

**Comment by garious** [comment]
@aeyakovenko, documented here: #1263.

**Comment by garious** [comment]
Corrected, thanks.

**Comment by garious** [comment]
@aeyakovenko, do you want to take a stab at a whole migration from `std::thread` to scoped threads in a separate PR? I think you'll view this PR differently in the context of those scoped threads in place.

**Pull request closed without merging**